ktor-server/ktor-server-plugins/ktor-server-sessions/jvm/src/io/ktor/server/sessions/KotlinxSessionSerializer.kt

```diff
@@ -0,0 +1,397 @@
+            StructureKind.MAP -> currentMap = mutableMapOf()
+            StructureKind.CLASS, PolymorphicKind.SEALED -> {
+                currentClassEncoder = SessionsBackwardCompatibleEncoder(serializersModule)
+                return currentClassEncoder!!
+            }
+            else -> throw IllegalArgumentException("Unsupported kind: ${descriptor.kind}")
+        }
+        return super.beginStructure(descriptor)
+    }
+
+    override fun endStructure(descriptor: SerialDescriptor) {
+        if (currentClassEncoder != null) {
+            val encoded = currentClassEncoder!!.result()
+            parametersBuilder.append(nextElementName, "##$encoded")
+            currentClassEncoder = null
+        } else if (currentList != null) {
+            val encoded = currentList!!
+                .joinToString("&") { it.encodeURLQueryComponent() }
+                .encodeURLQueryComponent()
+            parametersBuilder.append(nextElementName, "#cl$encoded")
+            currentList = null
+        } else if (currentMap != null) {
+            val encoded = currentMap!!
+                .map { (key, value) -> "${key.encodeURLQueryComponent()}=${value.encodeURLQueryComponent()}" }
+                .joinToString("&")
+                .encodeURLQueryComponent()
+            parametersBuilder.append(nextElementName, "#m$encoded")
+            currentMap = null
+        }
+        super.endStructure(descriptor)
+    }
+
+    override fun encodeValue(value: Any) {
+        val encoded = primitiveValue(value) ?: return
+        if (currentList != null) {
+            currentList!!.add(encoded)
+            return
+        } else if (currentMap != null) {
+            if (mapKey != null) {
+                currentMap!![mapKey!!] = encoded
+                mapKey = null
+            } else {
+                mapKey = encoded
+            }
+            return
+        }
+        parametersBuilder.append(nextElementName, encoded)
+    }
+
+    override fun encodeElement(descriptor: SerialDescriptor, index: Int): Boolean {
+        if (descriptor.kind != StructureKind.LIST && descriptor.kind != StructureKind.MAP) {
+            nextElementName = descriptor.getElementName(index)
+        }
+        return true
+    }
+
+    override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) {
+        encodeValue(enumDescriptor.getElementName(index))
+    }
+
+    override fun encodeNull() {
+        parametersBuilder.append(nextElementName, "#n")
+    }
+
+    private fun primitiveValue(value: Any) = when (value) {
+        is Int -> "#i$value"
+        is Long -> "#l$value"
+        is Float -> "#f$value"
+        is Double -> "#f$value"
+        is Boolean -> "#bo${value.toString().first()}"
+        is Char -> "#ch$value"
+        is String -> "#s$value"
+        is Enum<*> -> "#s${value.name}"
+        else -> null
+    }
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+private class SessionsBackwardCompatibleDecoder(
```

Comment:
Please consider splitting this file into multiple shorts. It's hard to maintain and review